File: app/local_plugins/devices/device_jieli/android/src/main/kotlin/com/jielihome/jielihome/feature/translation/mode/StereoCallTranslationModeHandler.kt
package com.jielihome.jielihome.feature.translation.mode

import android.content.Context
import com.jieli.bluetooth.bean.translation.AudioData
import com.jieli.bluetooth.bean.translation.TranslationMode
import com.jieli.bluetooth.constant.Constants
import com.jieli.bluetooth.impl.JL_BluetoothManager
import com.jielihome.jielihome.audio.PcmKit
import com.jielihome.jielihome.feature.ConnectFeature
import com.jielihome.jielihome.feature.translation.AudioFormat
import com.jielihome.jielihome.feature.translation.BaseTranslationModeHandler
import com.jielihome.jielihome.feature.translation.TranslationAudioBridge
import com.jielihome.jielihome.feature.translation.TranslationModeIds
import com.jielihome.jielihome.feature.translation.TranslationStreams
import com.jielihome.jielihome.feature.translation.runtime.RcspTranslationRuntime
import java.io.File

/**
 * MODE_CALL_TRANSLATION_WITH_STEREO —— 立体声通话翻译。
 * SDK 上来一路立体声 PCM（左=上行，右=下行），软件分离后等价于普通通话翻译两路。
 */
class StereoCallTranslationModeHandler(
    private val context: Context,
    btManager: JL_BluetoothManager,
    private val connectFeature: ConnectFeature,
    bridge: TranslationAudioBridge,
) : BaseTranslationModeHandler(btManager, bridge) {

    override val modeId = TranslationModeIds.MODE_CALL_TRANSLATION_WITH_STEREO
    override val inputStreams = listOf(TranslationStreams.IN_UPLINK, TranslationStreams.IN_DOWNLINK)
    override val outputStreams = listOf(TranslationStreams.OUT_UPLINK, TranslationStreams.OUT_DOWNLINK)

    private var runtime: RcspTranslationRuntime? = null

    override fun start(args: Map<String, Any?>) {
        if (working) return
        val address = args["address"] as? String
            ?: connectFeature.connectedDevice()?.address
        val device = address?.let { connectFeature.deviceByAddress(it) }
            ?: connectFeature.connectedDevice()
        if (device == null) {
            emitError(-200, "no connected device")
            return
        }
        val sampleRate = (args["sampleRate"] as? Int) ?: 16000
        val audioType = parseAudioType(args["audioType"]) ?: Constants.AUDIO_TYPE_OPUS
        // 与 demo 对齐：默认 ALWAYS_RECORDING，进 mode 即持续上推 PCM；
        // AUTO 会在没有真实通话事件时静默不出帧。
        val strategy = parseStrategy(args["strategy"])
            ?: TranslationMode.STRATEGY_DEVICE_ALWAYS_RECORDING
        val sdkMode = TranslationMode(
            TranslationMode.MODE_CALL_TRANSLATION_WITH_STEREO,
            audioType,
            2,                                    // 立体声
            sampleRate,
        ).setRecordingStrategy(strategy)

        val rt = RcspTranslationRuntime(
            btManager = btManager,
            device = device,
            mode = sdkMode,
            tempDir = File(context.cacheDir, "jieli_translation_tts"),
            onPcm = { source, stereoPcm ->
                if (source != AudioData.SOURCE_E_SCO_MIX) return@RcspTranslationRuntime
                val (left, right) = PcmKit.splitStereo16(stereoPcm)
                val fmt = AudioFormat(sampleRate, 1, 16)
                pushFrame(TranslationStreams.IN_UPLINK, left, fmt)
                pushFrame(TranslationStreams.IN_DOWNLINK, right, fmt)
            },
            onError = { code, msg -> emitError(code, msg) }
        )
        rt.start().fold(
            {
                runtime = rt
                working = true
                emitLog("StereoCallTranslation start device=${device.address} audioType=$audioType sampleRate=$sampleRate")
            },
            { emitError(-201, it.message); rt.stop() },
        )
    }

    private fun parseStrategy(raw: Any?): Int? = when (raw) {
        null -> null
        is Int -> raw
        is Number -> raw.toInt()
        is String -> when (raw.lowercase()) {
            "custom" -> TranslationMode.STRATEGY_CUSTOM_RECORDING
            "always" -> TranslationMode.STRATEGY_DEVICE_ALWAYS_RECORDING
            "auto" -> TranslationMode.STRATEGY_DEVICE_AUTO_RECORDING
            else -> raw.toIntOrNull()
        }
        else -> null
    }

    private fun parseAudioType(raw: Any?): Int? = when (raw) {
        null -> null
        is Int -> raw
        is Number -> raw.toInt()
        is String -> when (raw.lowercase()) {
            "opus" -> Constants.AUDIO_TYPE_OPUS
            "pcm" -> Constants.AUDIO_TYPE_PCM
            "speex" -> Constants.AUDIO_TYPE_SPEEX
            "msbc", "m_sbc" -> Constants.AUDIO_TYPE_M_SBC
            "jla_v2", "jlav2" -> Constants.AUDIO_TYPE_JLA_V2
            else -> raw.toIntOrNull()
        }
        else -> null
    }

    override fun stop() {
        if (!working) return
        runtime?.stop()
        runtime = null
        working = false
        emitLog("StereoCallTranslation stop")
    }

    override fun onTranslatedAudio(
        outputStreamId: String,
        pcm: ByteArray,
        format: AudioFormat,
        isFinal: Boolean,
    ): Boolean {
        if (outputStreamId != TranslationStreams.OUT_UPLINK &&
            outputStreamId != TranslationStreams.OUT_DOWNLINK) return false
        return runtime?.feedTtsPcm(outputStreamId, pcm, isFinal) ?: false
    }
}

File: apps/client/local_plugins/devices/device_jieli/ios/device_jieli/Sources/device_jieli/feature/translation/mode/StereoCallTranslationModeHandler.swift
import Foundation
import JL_BLEKit

/// MODE_CALL_TRANSLATION_WITH_STEREO（mode=6）—— 立体声通话翻译。
///
/// 与 Android `StereoCallTranslationModeHandler` 对齐：
///   - 耳机一路推 SOURCE_E_SCO_MIX 立体声 OPUS：L=本端 uplink, R=对端 downlink
///   - 软件分声道后等价于 [CallTranslationModeHandler]
public final class StereoCallTranslationModeHandler: BaseTranslationModeHandler {

    private var decoder: OpusStreamDecoder?
    private var encoderUplink: OpusStreamEncoder?
    private var encoderDownlink: OpusStreamEncoder?
    private var sampleRate = 16000
    private var audioType: JL_SpeakDataType = .OPUS

    public init(server: JieliHomeServer) {
        super.init(
            server: server,
            modeId: TranslationModeIds.MODE_CALL_TRANSLATION_WITH_STEREO,
            inputs: [TranslationStreams.IN_UPLINK, TranslationStreams.IN_DOWNLINK],
            outputs: [TranslationStreams.OUT_UPLINK, TranslationStreams.OUT_DOWNLINK]
        )
    }

    public override func start(args: [String: Any], completion: @escaping (Bool, String?) -> Void) {
        guard let server = server else { completion(false, "server uninitialized"); return }
        if working { completion(true, nil); return }
        let address = args["address"] as? String
        let session: TranslationSession? = address.flatMap { server.translationSession(for: $0) } ?? server.currentTranslationSession()
        guard let s = session else {
            emitError(-200, "no connected device")
            completion(false, "no connected device"); return
        }
        let sr = (args["sampleRate"] as? NSNumber)?.intValue ?? 16000
        let aType = Self.parseAudioType(args["audioType"]) ?? .OPUS
        let recordType = Self.parseRecordType(args["strategy"], default: .byDevice)
        sampleRate = sr
        audioType = aType
        self.session = s

        if aType == .OPUS {
            decoder = OpusStreamDecoder(channels: 2, dataSize: 80, sampleRate: sr) { [weak self] stereoPcm in
                guard let self = self else { return }
                let (l, r) = PcmKit.splitStereo16(stereoPcm)
                let fmt = AudioFormat(sampleRate: sr, channels: 1, bitsPerSample: 16)
                self.pushFrame(streamId: TranslationStreams.IN_UPLINK, pcm: l, format: fmt)
                self.pushFrame(streamId: TranslationStreams.IN_DOWNLINK, pcm: r, format: fmt)
            } onError: { [weak self] code, msg in
                self?.emitError(code, "opus stereo decode: \(msg ?? "")")
            }
            encoderUplink = OpusStreamEncoder(channels: 1, sampleRate: sr) { [weak self] opus in
                self?.writeBack(opus, source: .typeESCOUp)
            }
            encoderDownlink = OpusStreamEncoder(channels: 1, sampleRate: sr) { [weak self] opus in
                self?.writeBack(opus, source: .typeESCODown)
            }
        }

        s.acquire(owner: self) { [weak self] tm, errMsg in
            guard let self = self else { return }
            guard let tm = tm else {
                self.emitError(-201, errMsg)
                completion(false, errMsg); return
            }
            tm.recordtype = recordType
            let mode = Self.buildMode(modeType: .callTranslateStereo, audioType: aType, channel: 2, sampleRate: sr)
            self.emitLog("StereoCallTranslation start audioType=\(aType.rawValue) sr=\(sr) recordType=\(recordType.rawValue)")
            tm.trStartTranslate(mode) { [weak self] status, err in
                guard let self = self else { return }
                if status == .success || status == .sameMode {
                    self.working = true
                    completion(true, nil)
                } else {
                    self.emitError(Int(status.rawValue), err?.localizedDescription ?? "trStartTranslateMode failed")
                    self.stop()
                    completion(false, "tm status=\(status.rawValue)")
                }
            }
        }
    }

    public override func stop() {
        guard let s = session, working else { working = false; return }
        working = false
        s.release(owner: self)
        decoder?.release(); decoder = nil
        encoderUplink?.release(); encoderUplink = nil
        encoderDownlink?.release(); encoderDownlink = nil
        emitLog("StereoCallTranslation stop")
    }

    public override func onReceiveAudio(_ audio: JLTranslateAudio) {
        guard working else { return }
        guard audio.sourceType == .typeESCOMax else { return }  // SOURCE_E_SCO_MIX
        switch audio.audioType {
        case .OPUS: decoder?.feed(audio.data)
        case .PCM:
            let (l, r) = PcmKit.splitStereo16(audio.data)
            let fmt = AudioFormat(sampleRate: sampleRate, channels: 1, bitsPerSample: 16)
            pushFrame(streamId: TranslationStreams.IN_UPLINK, pcm: l, format: fmt)
            pushFrame(streamId: TranslationStreams.IN_DOWNLINK, pcm: r, format: fmt)
        default: break
        }
    }

    public override func onTranslatedAudio(streamId: String, pcm: Data, format: AudioFormat, isFinal: Bool) -> Bool {
        guard working else { return false }
        let encoder: OpusStreamEncoder?
        let source: JLTranslateAudioSourceType
        switch streamId {
        case TranslationStreams.OUT_UPLINK: encoder = encoderUplink; source = .typeESCOUp
        case TranslationStreams.OUT_DOWNLINK: encoder = encoderDownlink; source = .typeESCODown
        default: return false
        }
        if audioType == .OPUS {
            let frameBytes = 640
            var ofs = 0
            while ofs < pcm.count {
                let len = min(frameBytes, pcm.count - ofs)
                encoder?.feed(pcm.subdata(in: ofs..<(ofs + len)))
                ofs += len
            }
        } else {
            writeBack(pcm, source: source, audioType: .PCM)
        }
        return true
    }

    private func writeBack(_ data: Data, source: JLTranslateAudioSourceType, audioType: JL_SpeakDataType? = nil) {
        guard let tm = session?.currentManager() else { return }
        let audio = JLTranslateAudio()
        audio.sourceType = source
        audio.audioType = audioType ?? self.audioType
        audio.data = data
        audio.len = Int32(data.count)
        audio.count = 1
        tm.trWriteAudioV2(audio, translate: data)
    }
}
